healthcheck/healthcheck.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2021 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package healthcheck tests and communicates the health of the Cloud SQL Auth proxy.
+package healthcheck
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"sync/atomic"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
+)
+
+const (
+	livenessPath = "/liveness"
+	readinessPath = "/readiness"
+	portNum = ":8080" // TODO(monazhn): Think about a good port number.
+)
+
+// Locks to synchronize reads and writes for each HC boolean. Separate locks
+// for each boolean is ideal so more than one boolean can be read/written
+// simultaneously.
+var (
+	readinessL = &sync.Mutex{}
+	livenessL  = &sync.Mutex{}
+	startupL   = &sync.Mutex{}
+)
+
+// HC is a type used to implement health checks for the proxy.
+type HC struct {
+	// live being true means the proxy is running; in the case of the proxy
+	// being unexpectedly terminated, we should (re)start the proxy.
+	// live is related to Kubernetes liveness probing.
+	live  bool
+	// ready being true means the proxy is ready to serve new traffic; in the 
+	// case that ready is false, we should wait to send new traffic to the
+	// proxy. The value of ready determines the success or failure of 
+	// Kubernetes readiness probing.
+	ready bool
+	// started is a flag used to support readiness probing and should not be
+	// confused for affecting startup probing. When started becomes true, the
+	// proxy is done starting up.
+	started bool
+	// srv is a pointer to the HTTP server used to communicated proxy health.
+	srv *http.Server
+}
+
+// NewHealthCheck initializes a HC object and exposes HTTP endpoints used to
+// communicate proxy health.
+func NewHealthCheck(proxyClient *proxy.Client) *HC {
+	srv := &http.Server{
+		Addr: portNum,
+	}
+
+	hc := &HC{
+		live: true,
+		srv:  srv,
+	}
+
+	// Handlers used to set up HTTP endpoints.
+	http.HandleFunc(readinessPath, func(w http.ResponseWriter, _ *http.Request) {
+		readinessL.Lock()
+		hc.ready = readinessTest(proxyClient, hc)
```

Comment:
You can just use the local variable for the return value.  Only lock when you need to change the state of the global variable.  Make your critical section small.  like this:

res := readinessAPI(...);   // res is a local variable inside the thread/routine
lock
hc.read = res.    // hc.read is the global variable that the lock is use to protect
unlokc

Reply:
Concurrent programming is hard even for experienced engineers. I think in this project the initial goal can be just make the logic right, no deadlock or starving issues.  You can optimize the code later. You can even try async (promise/future) handlers if the test function is heavy.

Reply:
We decided for now that we'll run `readinessTest` each time and won't save the result, so this gets rid of some of the locks. I'll keep in mind the local variable and async handlers for the future. Thanks!